codespan-reporting/src/term/renderer.rs

```diff
@@ -266,12 +266,46 @@ impl<'writer, 'config> Renderer<'writer, 'config> {
 
             // Write source text
             write!(self, " ")?;
+            let mut in_primary = false;
             for (metrics, ch) in self.char_metrics(source.char_indices()) {
+                let column_range = metrics.byte_index..(metrics.byte_index + ch.len_utf8());
+
+                // Check if we are overlapping a primary label
+                let is_primary = single_labels.iter().any(|(label_style, range, _)| {
+                    *label_style == LabelStyle::Primary && is_overlapping(range, &column_range)
+                }) || multi_labels.iter().any(|(_, label)| match label {
+                    MultiLabel::Top(label_style, range) => {
+                        *label_style == LabelStyle::Primary && column_range.contains(&range.end)
+                    }
+                    MultiLabel::TopLeft(label_style) | MultiLabel::Left(label_style) => {
+                        *label_style == LabelStyle::Primary
+                    }
+                    MultiLabel::Bottom(label_style, range, _) => {
+                        *label_style == LabelStyle::Primary && column_range.contains(&range.end)
+                    }
+                });
+
+                // Set the source color if we are in a primary label
+                match (is_primary, in_primary) {
+                    (true, true) | (false, false) => {}
+                    (true, false) => {
+                        self.set_color(self.styles().label(severity, LabelStyle::Primary))?;
+                        in_primary = true;
+                    }
+                    (false, true) => {
+                        self.reset()?;
+                        in_primary = false;
+                    }
+                }
+
                 match ch {
                     '\t' => (0..metrics.unicode_width).try_for_each(|_| write!(self, " "))?,
                     _ => write!(self, "{}", ch)?,
                 }
             }
+            if in_primary {
+                self.reset()?;
+            }
```

Comment:
This whole section is rather ugly 😓 - not sure how to improve it.

Reply:
I am guessing a bit, but maybe it is better to keep a stack of styles while rendering and then just push/pop as different styles are applied.

Reply:
Yeah, perhaps I could have some state in the renderer that handles this. Good idea! I managed to clean it up a _bit_ though.